phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java

```diff
@@ -178,8 +178,11 @@ private static void throwIfScanOutOfRegion(Scan scan, Region region) throws DoNo
             byte[] expectedUpperRegionKey =
                     scan.getAttribute(EXPECTED_UPPER_REGION_KEY) == null ? scan.getStopRow() : scan
                             .getAttribute(EXPECTED_UPPER_REGION_KEY);
-            isStaleRegionBoundaries = expectedUpperRegionKey != null &&
-                    Bytes.compareTo(upperExclusiveRegionKey, expectedUpperRegionKey) != 0;
+
```

Comment:
The following is the tricky part. Previous in order to detect a stale region boundary cache we only looked at the upper region bound. That work fine as long as do not have an open scanner that has already reached the upper half of the region when we split. When the scan restarts after the split finishes, the scan restarts in the new upper daugther region and the upper might be the same (it definitely is when this was the last region of the table).

That always was a problem but since we're starting with larger number of the viewIndexId this is just more likely. My solution is to use the Scan's expected start key and compare that. For everything I tested that does the right thing.

Reply:
The main part I could use a double-check on... Did I understand the purpose of SCAN_ACTUAL_START_ROW correctly?
From looking at the code and going by the name it's the start row of the scan that we can use in Phoenix to restart the scan (and that can happen exactly when a split occurred and the scanner was restarted by HBase with the new region start key - assuming it fell into the second daughter region). If that is indeed the right interpretation that this should be correct.

Reply:
Can you please add a comment here with a short version of the explanation above?

Reply:
@lhofhansl I think that makes sense given what's happening when you get a [StaleRegionBoundaryCacheException](https://github.com/apache/phoenix/blob/f8cfa6b8b9f63ab080cdafa486bb4aabf043b1f8/phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java#L1315) and call BaseResultIterators#[recreateIterators](https://github.com/apache/phoenix/blob/f8cfa6b8b9f63ab080cdafa486bb4aabf043b1f8/phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java#L1405)

Reply:
Yeah. Good idea.